src/components/manager.ts

```diff
@@ -200,155 +231,139 @@ export class Manager {
-                    this.findAdditionalDependentFilesFromFls(filePath)
-                }
-                this.extension.logger.addLogMessage(`File watcher: responding to change in ${filePath}`)
-                const configuration = vscode.workspace.getConfiguration('latex-workshop')
-                if (configuration.get('latex.autoBuild.run') as string !== 'onFileChange') {
-                    return
-                }
-                if (this.extension.builder.disableBuildAfterSave) {
-                    this.extension.logger.addLogMessage('Auto Build Run is temporarily disabled during a second.')
-                    return
-                }
-                this.extension.logger.addLogMessage(`${filePath} changed. Auto build project.`)
-                if (this.localRootFile && configuration.get('latex.rootFile.useSubFile')) {
-                    this.extension.commander.build(true, this.localRootFile)
-                } else {
-                    this.extension.commander.build(true, rootFile)
-                }
-            })
-            this.fileWatcher.on('unlink', async (filePath: string) => {
-                this.extension.logger.addLogMessage(`File watcher: ${filePath} deleted.`)
-                this.fileWatcher.unwatch(filePath)
-                this.filesWatched.splice(this.filesWatched.indexOf(filePath), 1)
-                this.clearTexFileTree(filePath)
-                if (filePath === rootFile) {
-                    this.extension.logger.addLogMessage(`Deleted ${filePath} was root - triggering root search`)
-                    await this.findRoot()
-                }
-            })
-            this.clearTexFileTree(rootFile)
-            this.findDependentFiles(rootFile)
-            this.findAdditionalDependentFilesFromFls(rootFile)
-        }
+    /* This function is called when a root file is found or a watched file is
+       changed (in vscode or externally). It searches the subfiles, including
+       \input siblings, bib files, and related fls file to construct a file
+       dependency data structure in `this.cachedContent`. Noted that only the
+       provided `file` is re-parsed, together with any new files that were not
+       previously watched/considered. Since this function is called upon content
+       changes, this lazy loading should be fine. */
+       async parseFileAndSubs(file: string, onChange: boolean = false) {
+        this.extension.logger.addLogMessage(`Parsing ${file}`)
+        if (this.filesWatched.indexOf(file) < 0) {
+            // The file is first time considered by the extension.
+            this.extension.logger.addLogMessage(`Adding ${file} to file watcher.`)
+            this.fileWatcher.add(file)
+            this.filesWatched.push(file)
+            // So update completer w.r.t. this file. Subsequent updates will be
+            // handled in 1) onDidChangeTextDocument (in vscode) and 2) file
+            // watcher (changed externally)
```

Comment:
A long time ago, we had two handlers as you describe: one for changes inside vscode and one for external change. Then, we only kept the external one, which actually watches when a file changes on disk and therefore responds when a file is saved inside vscode. I am wondering if you really need to keep both. If I remember well, `onDidChangeTextDocument` is called after every keystroke or so.